README.md
# cyblock

A simple state management using dart stream.

## What is Cyblock?

Cyblock is helper to use dart stream as state management, we can easily update data to listener with stream.
it inject an event to run callback for mapping event to state.

## Usage

We can create a class that extends Cybloc, there's 2 parameter. the first parameter is data type for stream, and the
second parameter is data type for event.

```dart
/// Create cybloc
/// The example is counter, there's int and CounterEvent
class CounterCyblock extends Cyblock<int, CounterEvent> {
  static final CounterCyblock _singleton = CounterCyblock._();
  CounterCyblock._() {
    initialState(0);
  }

  static CounterCyblock get instance => _singleton;

  @override
  void mapEventToState(CounterEvent event) {
    if(event is IncrementEvent) {
      emit(state+1);
    } else if(event is DecrementEvent) {
      emit(state-1);
    }
  }
}
```

and the CounterEvent is

```dart
/// Events
abstract class CounterEvent {}

class IncrementEvent extends CounterEvent {}

class DecrementEvent extends CounterEvent {}
```

so basically, we just inject a `CounterEvent` to cyblock and run `mapEventToState` callback to handle the logic.
we can use `emit(value)` to update the state, and we can get the current state in `state`

## Implementing on UI

The usage if simply, we can use `StreamBuilder` widget to build stream to widget

```dart
StreamBuilder<int>(
    stream: CounterCyblock.instance.stream,
    builder: (context, snapshot) {
        if(snapshot.connectionState == ConnectionState.waiting) {
            CounterCyblock.instance.getState();
            return Text("Not inserted yet");
        } else {
            return Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
                Text("The button is pressed :"),
                Text(snapshot.data.toString(), style: TextStyle(fontSize: 25.0, fontWeight: FontWeight.bold)),
            ],
            );
        }
    },
),
```

use `stream` to get stream on cyblock controller.
and we can inject event with `insertEvent(event)`.

```dart
floatingActionButton: Row(
    mainAxisAlignment: MainAxisAlignment.end,
    children: [
        FloatingActionButton(
            heroTag: "decrement",
            child: Icon(Icons.remove),
            onPressed: () {
                CounterCyblock.instance.insertEvent(DecrementEvent());
            },
        ),
        SizedBox(width: 10.0,),
        FloatingActionButton(
            heroTag: "increment",
            child: Icon(Icons.add),
            onPressed: () {
                CounterCyblock.instance.insertEvent(IncrementEvent());
            },
        )
    ],
),
```

## Disclaimer

Im still new on dart and flutter, and this is just my experiment for flutter state management.
feel free to fork and contribute :)


CHANGELOG.md
## [0.0.1+4] - 2020-11-29

* Sync-up

## [0.0.1+3] - 2020-10-10

* Shorten description, too long :)

## [0.0.1+2] - 2020-10-10

* Longer description
* Format core with dartfmt
* Update dependencies

## [0.0.1+1] - 2020-10-10

* Updating README.md

## [0.0.1] - 2020-10-10

* Initial push
* Adding example



test/widget_test.dart
// // This is a basic Flutter widget test.
// //
// // To perform an interaction with a widget in your test, use the WidgetTester
// // utility that Flutter provides. For example, you can send tap and scroll
// // gestures. You can also use WidgetTester to find child widgets in the widget
// // tree, read text, and verify that the values of widget properties are correct.

// import 'package:flutter/material.dart';
// import 'package:flutter_test/flutter_test.dart';

// import 'package:cyblock/main.dart';

// void main() {
//   testWidgets('Counter increments smoke test', (WidgetTester tester) async {
//     // Build our app and trigger a frame.
//     await tester.pumpWidget(MyApp());

//     // Verify that our counter starts at 0.
//     expect(find.text('0'), findsOneWidget);
//     expect(find.text('1'), findsNothing);

//     // Tap the '+' icon and trigger a frame.
//     await tester.tap(find.byIcon(Icons.add));
//     await tester.pump();

//     // Verify that our counter has incremented.
//     expect(find.text('0'), findsNothing);
//     expect(find.text('1'), findsOneWidget);
//   });
// }


test/cyblock_test.dart
// import 'package:flutter_test/flutter_test.dart';

// import 'package:cyblock/cyblock.dart';

// void main() {
//   test('adds one to input values', () {
//     final calculator = Calculator();
//     expect(calculator.addOne(2), 3);
//     expect(calculator.addOne(-7), -6);
//     expect(calculator.addOne(0), 1);
//     expect(() => calculator.addOne(null), throwsNoSuchMethodError);
//   });
// }


lib/services/core.dart
part of cyblock;

abstract class Cyblock<T, S> {
  /// Saved state during emit
  T state;

  /// Setup initial state function
  void initialState(T initState) {
    emit(initState);
  }

  /// Controller for main state stream
  final _streamController = new StreamController<T>.broadcast();

  /// for inputing state
  /// but you can use emit instead of using [_inStream] manually
  StreamSink<T> get _inStream => _streamController.sink;

  /// If you want to stream on the cybloc you can use [stream]
  ///
  /// [SomeCyblocClass.stream]
  Stream<T> get stream => _streamController.stream;
  final _eventController = new StreamController<S>.broadcast();
  StreamSink<S> get _eventListen => _eventController.sink;

  /// Use [emit] to update state
  void emit(T data) {
    state = data;
    _inStream.add(data);
    if(data == null) {
      _inBuilderStream.add(null);
    } else {
      _inBuilderStream.add(CyblockSuccessState<T>(
        data: data,
        dateTime: DateTime.now()
      ));
    }
  }

  void throwError(Exception e) {
    _inBuilderStream.add(CyblockFailureState<T>(
      error: e,
      data: state??null
    ));
  }

  /// Disposing Stream Controller
  void dispose() {
    _streamController?.close();
    _eventController?.close();
    _builderController?.close();
  }

  /// Use [insertEvent] to insert event
  void insertEvent(S event) {
    _eventListen.add(event);
  }

  /// mapping event to state
  void mapEventToState(S event);

  /// Function to get current state
  getState() {
    emit(state);
  }

  /// Controller for [CyblockBuilder]
  final _builderController = new StreamController<CyblockStates>.broadcast();
  
  /// Stream pipe for builder
  Stream<CyblockStates> get builderStream => _builderController.stream;
  
  /// Input state for builder
  StreamSink<CyblockStates> get _inBuilderStream => _builderController.sink;

  /// to construct the initial value, use `Cyblock() : super(initialValue);`
  Cyblock(T initState) {
    state = initState; 
    _eventController.stream.listen(mapEventToState);
  }
}

abstract class Cybit<T> {
  /// Saved state during emit
  T state;

  /// Setup initial state function
  void initialState(T initState) {
    emit(initState);
  }

  /// Controller for main state stream
  final _streamController = new StreamController<T>.broadcast();

  /// for inputing state
  /// but you can use emit instead of using [_inStream] manually
  StreamSink<T> get _inStream => _streamController.sink;

  /// Use [emit] to update state
  void emit(T data) {
    state = data;
    _inStream.add(data);
  
    _inBuilderStream.add(CyblockSuccessState<T>(
      data: data,
      dateTime: DateTime.now()
    ));
  }

  void throwError(dynamic e) {
    _inBuilderStream.add(CyblockFailureState<T>(
      error: e,
      data: state
    ));
  }

  /// Disposing Stream Controller
  void dispose() {
    _streamController?.close();
    _builderController?.close();
  }

  /// Function to get current state
  getState() {
    _inStream.add(state);
  
    _inBuilderStream.add(CyblockSuccessState<T>(
      data: state,
      dateTime: DateTime.now()
    ));
  }

  /// Controller for [CyblockBuilder]
  final _builderController = new StreamController<CyblockStates>.broadcast();
  
  /// Stream pipe for builder
  Stream<CyblockStates> get builderStream => _builderController.stream;
  
  /// Input state for builder
  StreamSink<CyblockStates> get _inBuilderStream => _builderController.sink;

  /// to construct the initial value, use `Cyblock() : super(initialValue);`
  Cybit(T initState) {
    initialState(initState);
  }
}


lib/widget/cyblock_builder.dart
part of cyblock;

/// T is for state type [int, String or custom data model]
class CyblockBuilder<T> extends StatelessWidget {
  /// Use the cyblock for this parameter
  final Cyblock cyblock;

  /// On Function Sucess, have 2 parameters [BuildContext] and [CyblockSuccessState], return [Widget]
  final OnSuccess<T> onSuccess;

  /// On Function Failure, have 2 parameteres [BuildContext] and [CyblockFailureState], return [Widget]
  final OnFailure<T> onFailure;

  /// On Loading, return [Widget]
  final OnLoading onLoading;

  /// initial function if there's no state
  final VoidCallback initial;

  CyblockBuilder({
    @required this.cyblock,
    @required this.onSuccess,
    @required this.onLoading,
    @required this.initial,
    this.onFailure,
  });

  /// Initial function
  onInitial() {
    cyblock.emit(null);
    initial();
  }
  
  @override
  Widget build(BuildContext context) {
    return StreamBuilder<CyblockStates>(
      stream: cyblock.builderStream,
      builder: (context, snapshot) {
        if(snapshot.connectionState == ConnectionState.waiting) {
          cyblock.getState();
          return onLoading(context);
        }

        if(snapshot.data == null && initial != null) {
          onInitial();
        }

        if(snapshot.data is CyblockSuccessState<T>) {
          return onSuccess(context, snapshot.data);
        } else if(snapshot.data is CyblockFailureState<T>) {
          return onFailure(context, snapshot.data, onInitial);
        }

        return onLoading(context);
      },
    );
  }
}

/// Typedef for Success
typedef Widget OnSuccess<T>(BuildContext context, CyblockSuccessState<T> successState);

/// Typedef for Failure
typedef Widget OnFailure<T>(BuildContext context, CyblockFailureState<T> failureState, VoidCallback reload);

/// Typedef for Loading
typedef Widget OnLoading(BuildContext context);

example/lib/view/with_builder.dart
import 'package:cyblock/cyblock.dart';
import 'package:flutter/material.dart';


class WithBuilderMainView extends StatefulWidget {
  @override
  _WithBuilderMainViewState createState() => _WithBuilderMainViewState();
}

class _WithBuilderMainViewState extends State<WithBuilderMainView> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        actions: [
          IconButton(
            icon: Icon(Icons.add),
            onPressed: () {
              CounterCyblockWithBuilder.instance.addWithError();
            },
          )
        ],
        title: Text("With Builder"),
      ),
      floatingActionButton: Row(
        mainAxisAlignment: MainAxisAlignment.end,
        children: [
          FloatingActionButton(
            child: Icon(Icons.add),
            heroTag: "increment",
            onPressed: () {
              CounterCyblockWithBuilder.instance.add(1);
            },
          ),
          FloatingActionButton(
            child: Icon(Icons.remove),
            heroTag: "decrement",
            onPressed: () {
              CounterCyblockWithBuilder.instance.remove(1);
            },
          )
        ],
      ),
      body: CyblockBuilder<int>(
        initial: CounterCyblockWithBuilder.instance.getState,
        cyblock: CounterCyblockWithBuilder.instance,
        onSuccess: (context, data) {
          return Center(
            child: Text(data.data.toString()),
          );
        },
        onLoading: (context) {
          return Center(
            child: CircularProgressIndicator(),
          );
        },
        onFailure: (context, error, callback) {
          return Center(
            child: FlatButton(
              child: Text("Reload"),
              onPressed: callback
            ),
          );
        },
      ),
    );
  }
}

class CounterCyblockWithBuilder extends Cyblock<int, String> {
  static final CounterCyblockWithBuilder _singleton = CounterCyblockWithBuilder._();

  static CounterCyblockWithBuilder get instance => _singleton;

  CounterCyblockWithBuilder._() : super(0);

  add(int count) {
    emit(state+count);
  }

  remove(int count) {
    emit(state-count);
  }

  addWithError() async {
    try {
      await Future.delayed(Duration(seconds: 1));
      throw "Error";
    } catch (e) {
      throwError(e);
    }
  }

  @override
  void mapEventToState(String event) {
    // TODO: implement mapEventToState
  }
}

example/lib/main.dart
import 'package:cyblock/cyblock.dart';
import 'package:flutter/material.dart';

import 'view/with_builder.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: "Example app",
      home: HomeMainView(),
    );
  }
}

class HomeMainView extends StatefulWidget {
  @override
  _HomeMainViewState createState() => _HomeMainViewState();
}

class _HomeMainViewState extends State<HomeMainView> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        actions: [
          IconButton(
            icon: Icon(Icons.navigate_before),
            onPressed: () {
              Navigator.push(context, MaterialPageRoute(
                builder: (context) => WithBuilderMainView()
              ));
            },
          )
        ],
        title: Text("Cyblock example"),
      ),
      body: SizedBox.expand(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            StreamBuilder<int>(
              stream: CounterCyblock.instance.stream,
              builder: (context, snapshot) {
                if(snapshot.connectionState == ConnectionState.waiting) {
                  CounterCyblock.instance.getState();
                  return Text("Not inserted yet");
                } else {
                  return Column(
                    mainAxisAlignment: MainAxisAlignment.center,
                    children: [
                      Text("The button is pressed :"),
                      Text(snapshot.data.toString(), style: TextStyle(fontSize: 25.0, fontWeight: FontWeight.bold)),
                    ],
                  );
                }
              },
            ),
            FlatButton(
              child: Text("Another page"),
              onPressed: () {
                Navigator.push(context, MaterialPageRoute(
                  builder: (_) => SecondPageMainView()
                ));
              },
            )
          ],
        ),
      ),
      floatingActionButton: Row(
        mainAxisAlignment: MainAxisAlignment.end,
        children: [
          FloatingActionButton(
            heroTag: "decrement",
            child: Icon(Icons.remove),
            onPressed: () {
              CounterCyblock.instance.insertEvent(DecrementEvent());
            },
          ),
          SizedBox(width: 10.0,),
          FloatingActionButton(
            heroTag: "increment",
            child: Icon(Icons.add),
            onPressed: () {
              CounterCyblock.instance.insertEvent(IncrementEvent());
            },
          )
        ],
      ),
    );
  }
}

class SecondPageMainView extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("Second page"),
      ),
      body: Center(
        child: StreamBuilder<int>(
          stream: CounterCyblock.instance.stream,
          builder: (context, snapshot) {
            if(snapshot.connectionState == ConnectionState.waiting) {
              CounterCyblock.instance.getState();
              return CircularProgressIndicator();
            } else {
              return Text(snapshot.data.toString());
            }
          },
        ),
      ),
      floatingActionButton: Row(
        mainAxisAlignment: MainAxisAlignment.end,
        children: [
          FloatingActionButton(
            heroTag: "decrement",
            child: Icon(Icons.remove),
            onPressed: () {
              CounterCyblock.instance.insertEvent(DecrementEvent());
            },
          ),
          SizedBox(width: 10.0,),
          FloatingActionButton(
            heroTag: "increment",
            child: Icon(Icons.add),
            onPressed: () {
              CounterCyblock.instance.insertEvent(IncrementEvent());
            },
          )
        ],
      ),
    );
  }
}

/// Create cybloc
class CounterCyblock extends Cyblock<int, CounterEvent> {
  static final CounterCyblock _singleton = CounterCyblock._();
  CounterCyblock._() : super(10);

  static CounterCyblock get instance => _singleton;

  @override
  void mapEventToState(CounterEvent event) {
    if(event is IncrementEvent) {
      emit(state+1);
    } else if(event is DecrementEvent) {
      emit(state-1);
    }
  }
}


/// Events
abstract class CounterEvent {}

class IncrementEvent extends CounterEvent {}

class DecrementEvent extends CounterEvent {}
